rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,556 @@
+    ) -> Self {
+        TokenSwapper {
+            graph,
+            input_mapping: mapping,
+            trials: trials.unwrap_or(4),
+            rng_seed: seed,
+            node_map: HashMap::with_capacity(graph.node_count()),
+            rev_node_map: HashMap::with_capacity(graph.node_count()),
+        }
+    }
+
+    fn map(&mut self) -> Vec<Swap> {
+        let num_nodes = self.graph.node_count();
+        let num_edges = self.graph.edge_count();
+
+        // Directed graph with nodes matching ``graph`` and
+        // edges for neighbors closer than nodes
+        let mut digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // Same as digraph with no self edges
+        let mut sub_digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // A list of nodes that are remaining to be tried
+        let mut todo_nodes = Vec::with_capacity(self.input_mapping.len());
+
+        let mut rng_seed: Pcg64 = match self.rng_seed {
+            Some(rng_seed) => Pcg64::seed_from_u64(rng_seed),
+            None => Pcg64::from_entropy(),
+        };
+        // Add nodes to the digraph/sub_digraph and build the node maps
+        for node in self.graph.node_identifiers() {
+            let index = digraph.add_node(());
+            sub_digraph.add_node(());
+            self.node_map.insert(node, index);
+            self.rev_node_map.insert(index, node);
+        }
+        // The input mapping in HashMap form using NodeIndex
+        let mut tokens: HashMap<NodeIndex, NodeIndex> = self
+            .input_mapping
+            .iter()
+            .map(|(k, v)| (self.node_map[k], self.node_map[v]))
+            .collect();
+        // todo_nodes are all the mapping entries where left != right
+        for (node, dest) in &tokens {
+            if node != dest {
+                todo_nodes.push(*node);
+            }
+        }
+        // Add initial edges to the digraph/sub_digraph
+        for node in self.graph.node_identifiers() {
+            self.add_token_edges(
+                self.node_map[&node],
+                &mut digraph,
+                &mut sub_digraph,
+                &mut tokens,
+            );
+        }
+        // Do the trials and build a results Vec
+        let mut trial_results: Vec<Vec<Swap>> = Vec::new();
+        for _ in 0..self.trials {
+            let results = self.trial_map(
+                &mut digraph.clone(),
+                &mut sub_digraph.clone(),
+                &mut tokens.clone(),
+                &mut todo_nodes.clone(),
+                &mut rng_seed,
+            );
+            trial_results.push(results);
+        }
+        // Build the first results Vec until a 0 len result is found
+        let mut first_results: Vec<Vec<Swap>> = Vec::new();
+        for result in trial_results {
+            if result.is_empty() {
+                first_results.push(result);
+                break;
+            }
+            first_results.push(result);
+        }
```

Comment:
I would integrate this with my iterator example above and make this just a filter function with a boolean, something like:

```rust
let mut empty_found = false;

...

.filter(|x| {
    if empty_found {
        None
    } else {
        if x.is_empty() {
            empty_found = true;
        }
        Some(x)
    }
})
.collect()
```

If you combine this with the iterator I suggested above this will remove an intermediate allocation and be much more efficient.